ui/src/shared/apis/queryCache.ts

```diff
@@ -0,0 +1,194 @@
+import {filterUnusedVarsBasedOnQuery} from 'src/shared/utils/filterUnusedVars'
+import {event} from 'src/cloud/utils/reporting'
+
+// Types
+import {RunQueryResult} from 'src/shared/apis/query'
+import {CancelBox} from 'src/types/promises'
+import {AppState, GetState, Variable} from 'src/types'
+import {RunQueryPromiseMutex} from './singleQuery'
+
+export const TIME_INVALIDATION = 1000 * 60 * 10 // 10 minutes
+
+const asSimplyKeyValueVariables = (vari: Variable) => {
+  return {
+    [vari.name]: vari.selected || [],
+  }
+}
+
+// Hashing function found here:
+// https://jsperf.com/hashcodelordvlad
+// Through this thread:
+// https://stackoverflow.com/questions/7616461/generate-a-hash-from-string-in-javascript
+const hashCode = (rawText: string): string => {
+  let hash = 0
+  if (!rawText) {
+    return `${hash}`
+  }
+  for (let i = 0; i < rawText.length; i++) {
+    hash = (hash << 5) - hash + rawText.charCodeAt(i)
+    hash |= 0 // Convert to 32bit integer
+  }
+  return `${hash}`
+}
+
+class QueryCache {
+  cache = {}
+
+  private cleanExpiredQueries = (): void => {
+    const now = Date.now()
+    for (const id in this.cache) {
+      // TODO(ariel): need to implement specific rules for custom time ranges
+      if (this.cache[id].isCustomTime) {
+        continue
+      }
+      if (now - this.cache[id].dateSet > TIME_INVALIDATION) {
+        this.resetCacheByID(id)
+      }
+    }
+  }
+
+  getFromCache = (
+    id: string,
+    hashedVariables: string
+  ): RunQueryResult | null => {
+    // no existing query match
+    if (!this.cache[id]) {
+      return null
+    }
+    // query match with no existing variable match
+    if (this.cache[id].hashedVariables !== hashedVariables) {
+      this.resetCacheByID(id)
+      return null
+    }
+    // query has been initialized but the result has not been set
+    if (this.cache[id].values === undefined) {
+      return null
+    }
+    // query & variable match with an expired result
+    if (Date.now() - this.cache[id].dateSet > TIME_INVALIDATION) {
+      this.resetCacheByID(id)
+      return null
+    }
+    event('query_cache_hit', {context: 'queryCache'})
+    return this.cache[id].values
+  }
+
+  initializeCacheByID = (
+    queryID: string,
+    hashedVariables: string,
+    isCustomTime: boolean = false
+  ) => {
+    if (this.cache[queryID]) {
+      return this.cache[queryID]
+    }
+    this.cache[queryID] = {
+      dateSet: Date.now(),
+      hashedVariables,
+      isCustomTime,
+      mutex: RunQueryPromiseMutex<RunQueryResult>(),
+    }
+    return this.cache[queryID]
+  }
+
+  resetCacheByID = (id: string): void => {
+    if (!this.cache[id]) {
+      return
+    }
+    delete this.cache[id]
+  }
+
+  resetCache = (): void => {
+    this.cache = {}
+  }
+
+  setCacheByID = (
+    queryID: string,
+    hashedVariables: string,
+    values: RunQueryResult
+  ): void => {
+    event('query_cache_miss', {context: 'queryCache'})
```

Comment:
what's the context for?

Reply:
🤷 I just went based off of what was going on in the other usages. Is there a doc somewhere on event reporting that we should stick to?

Reply:
we might want to add the `queryID` into that object so that we can separate parallelized queries